c/secp256k1_blake160_sighash_all.c

```diff
@@ -1,191 +1,180 @@
 
   /* Load args */
+  unsigned char script[SCRIPT_SIZE];
   len = SCRIPT_SIZE;
   ret = ckb_load_script(script, &len, 0);
   if (ret != CKB_SUCCESS) {
     return ERROR_SYSCALL;
   }
-  script_pos.ptr = (const uint8_t*)script;
+  if (len > SCRIPT_SIZE) {
+    return ERROR_SCRIPT_TOO_LONG;
+  }
+  mol_pos_t script_pos;
+  script_pos.ptr = (const uint8_t *)script;
   script_pos.size = len;
-  args_res = mol_cut(&script_pos, MOL_Script_args());
+
+  mol_read_res_t args_res = mol_cut(&script_pos, MOL_Script_args());
   if (args_res.code != 0) {
     return ERROR_ENCODING;
   }
-  bytes_res = mol_cut_bytes(&args_res.pos);
-  if (bytes_res.code != 0) {
+  mol_read_res_t args_bytes_res = mol_cut_bytes(&args_res.pos);
+  if (args_bytes_res.code != 0) {
     return ERROR_ENCODING;
-  } else if (bytes_res.pos.size == 0) {
-    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  } else if (args_bytes_res.pos.size != BLAKE160_SIZE) {
+    return ERROR_ARGUMENTS_LEN;
   }
 
-  /* Calculate sigs count and threshold */
-  sigs_cnt = bytes_res.pos.size / BLAKE160_SIZE;
-  if (bytes_res.pos.size % BLAKE160_SIZE == 0) {
-    threshold = sigs_cnt;
-  } else {
-    threshold = bytes_res.pos.ptr[sigs_cnt * BLAKE160_SIZE];
+  /* Check witnesses is less than or equals to inputs  */
+  ret = check_witnesses_len();
+  if (ret != CKB_SUCCESS) {
+    return ERROR_INVALID_WITNESSES_COUNT;
   }
 
-  if (threshold > sigs_cnt || threshold == 0) {
-    return ERROR_INVALID_THRESHOLD;
+  /* Load witness of first input */
+  unsigned char witness[MAX_WITNESS_SIZE];
+  uint64_t witness_len = MAX_WITNESS_SIZE;
+  ret = ckb_load_witness(witness, &witness_len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
   }
 
-  uint8_t used_signatures[sigs_cnt];
+  if (witness_len > MAX_WITNESS_SIZE) {
+    return ERROR_WITNESS_SIZE;
+  }
 
-  secp256k1_context context;
-  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
+  /* load signature */
+  mol_read_res_t lock_bytes_res;
+  ret = extract_witness_lock(witness, witness_len, &lock_bytes_res);
   if (ret != 0) {
-    return ret;
+    return ERROR_ENCODING;
   }
 
+  if (lock_bytes_res.pos.size != SIGNATURE_SIZE) {
+    return ERROR_ARGUMENTS_LEN;
+  }
+  memcpy(lock_bytes, lock_bytes_res.pos.ptr, lock_bytes_res.pos.size);
+
+  /* Load tx hash */
+  unsigned char tx_hash[BLAKE2B_BLOCK_SIZE];
   len = BLAKE2B_BLOCK_SIZE;
   ret = ckb_load_tx_hash(tx_hash, &len, 0);
   if (ret != CKB_SUCCESS) {
     return ERROR_SYSCALL;
   }
 
+  /* Prepare sign message */
+  unsigned char message[BLAKE2B_BLOCK_SIZE];
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+
+  /* Clear lock field to zero, then digest the first witness */
+  memset((void *)lock_bytes_res.pos.ptr, 0, lock_bytes_res.pos.size);
+  blake2b_update(&blake2b_ctx, (char *)&witness_len, sizeof(uint64_t));
```

Comment:
I wish 3rd-party developers can understand this: `(char *)&witness_len` (since little endian).